R/validate-config-utils.R

```diff
@@ -301,7 +301,112 @@ find_invalid_target_keys <- function(target_keys, model_task_grp) {
     stats::setNames(names(target_keys))
 }
 
+# Validate the range (minimum & maximum) of acceptable sample numbers for a
+# given modeling task group in a given round.
+# Returns NULL if not applicable or check passes and error df row if check fails.
+validate_mt_sample_range <- function(model_task_grp,
+                                     model_task_i,
+                                     round_i,
+                                     schema) {
+  sample_config <- purrr::pluck(
+    model_task_grp,
+    "output_type",
+    "sample",
+    "output_type_id_params"
+  )
+
+  if (is.null(sample_config)) {
+    return(NULL)
+  }
+
+  check <- sample_config$max_samples_per_task < sample_config$min_samples_per_task
+
+  if (check) {
+    error_row <- data.frame(
+      instancePath = glue::glue(
+        get_error_path(
+          schema,
+          "output_type/sample/output_type_id_params",
+          "instance"
+        )
+      ),
+      schemaPath = paste(
+        get_error_path(
+          schema,
+          "sample",
+          "schema"
+        ), "output_type_id_params",
+        sep = "/"
+      ),
+      keyword = "Sample number range",
+      message = glue::glue(
+        "min_samples_per_task must be less or equal to max_samples_per_task."
+      ),
+      schema = "",
+      data = glue::glue("min_samples_per_task: {sample_config$min_samples_per_task};
+            max_samples_per_task: {sample_config$max_samples_per_task}")
+    )
+    return(error_row)
+  }
+  return(NULL)
+}
+
+# Validate that compound_taskid_set values are valid task ids for a
+# given modeling task group in a given round.
+# Returns NULL if not applicable or check passes and error df row if check fails.
+validate_mt_sample_comp_tids <- function(model_task_grp,
+                                         model_task_i,
+                                         round_i,
+                                         schema) {
```

Comment:
Questions about what's going on here and in `check_compound_taskids_valid`:

- The idea is that `check_compound_taskids_valid` is used when creating a new config file, but this function is used when validating an existing config file?
- There may be value in standardizing on naming conventions:
    - it's not clear to me why we've used `validate` here and `check` in the name of the other function.  The `check` function is actually `abort`ing, while this one generates an error message in a data frame that is collected for later use?  Are there standard conventions for this kind of thing?  I went and looked at `checkmate` docs, but they don't have things starting with `validate`.
    - noting `comp_tids` here vs. `compound_taskids` in the other function name
- would there be any value or possibility to refactor the check logic here and in that other function?  It's not much code and it's not exactly identical, so maybe not.  But it is quite similar.

Reply:
So there's two places that dynamic validations that can't be encoded in our schema need to be performed:
1. when validating a `tasks.json` file (`validate_mt_sample_comp_tids()`). I used `validate` here because it is part of the config validation process which itself is based on the concept of using json schema to validate json data, powered by the package `jsonvalidate`. The output of the `validate` function conform to the output of the `jsonvalidate::json_validate()`, specifically the `errors` attribute tbl which can then be passed to `view_config_val_errors()`.
2. when checking inputs to functions that create config programmatically.  Because we are checking inputs to a function effectively, here I opted for check. e.g. a la `rlang:: check_rquired()` etc which indeed is designed to throw an error if the condition is not met. 

I believe different languages have related but slightly different conventions (e.g. python uses `assert` in testing while we use `expect`) and I don't know of any resource is R that dictate the use of specific verbs for specific situations (I believe the checkmate convention is just their own internal convention). As such I've just tried to keep these internally consistent to separate functions that are used for validation of json vs functions used to check inputs to functions generating config. I've also kept such functions separate because although they might be checking the same thing, the inputs and outputs/side effects of each function are so different and take up so much of the logic in each function that trying to shoehorn them into a single function didn't make sense to me.

Reply:
thanks! all sounds good. the one remaining question then is about `comp_tids` vs. `compound_taskids`

Reply:
Fixed in 33a4859da417350731fc5920f1925b2f0354058a